wit/network.wit

```diff
@@ -13,11 +13,77 @@ default interface network {
 	drop-network: func(this: network)
 
 
-
+	/// Error codes.
+	/// 
+	/// In theory, every API can return any error code.
+	/// In practice, API's typically only return the errors documented per API
+	/// combined with a couple of errors that are always possible:
+	/// - `unknown`
+	/// - `access-denied`
+	/// - `not-supported`
+	/// - `resource-limit-reached`
+	/// - `out-of-memory`
+	/// 
+	/// See each individual API for what the POSIX equivalents are. They sometimes differ per API.
 	enum error {
+		// ### GENERAL ERRORS ###
+
+		/// Unknown error
 		unknown,
-		again,
-		// TODO ...
+
+		/// POSIX equivalent: EACCES, EPERM
+		access-denied,
+
+		/// POSIX equivalent: EOPNOTSUPP
+		not-supported,
+
+		/// POSIX equivalent: EMFILE, ENFILE, ENOMEM, ENOBUFS, EAI_MEMORY
+		resource-limit-reached,
+
+		/// POSIX equivalent: ENOMEM, ENOBUFS, EAI_MEMORY
+		out-of-memory,
+
+		timeout,
+		concurrency-conflict,
+
+
+		// ### IP ERRORS ###
+		address-family-not-supported,
+		address-family-mismatch,
+		invalid-remote-address,
+		ipv4-only-operation,
+		ipv6-only-operation,
+
+
+
+		// ### TCP & UDP SOCKET ERRORS ###
+		already-attached,
+		already-bound,
+		already-connected,
+		not-bound,
+		not-connected,
+		address-not-bindable,
+		address-in-use,
+		ephemeral-ports-exhausted,
+		remote-unreachable,
+		
+
+		// ### TCP SOCKET ERRORS ###
+		already-listening,
+		not-listening,
+		connection-refused,
+		connection-reset,
```

Comment:
Should there be a code corresponding to `ECONNABORTED`?

Reply:
AFAIK, `ECONNABORTED` is only returned by `accept`. Which, in my opinion should not be forwarded into WASI land. This is documented further along in the document:

> Host implementations must skip over transient errors returned by the native accept syscall.

Reply:
This is a little surprising. Indeed many users will want to just ignore `ECONNABORTED` and keep waiting for the next connection, but some users may want to know that the failure occurred so that they can log it. Large numbers of reset connections could be a sign of a network problem or unexpected client behavior. Some users may expect specific connections and want to know that a connection was attempted and reset so that they don't continue to wait for it.

Reply:
The choice of skipping transient errors stems from before the Preview2 PR. Before that, `accept` returned a single `stream` of TCP sockets, that would only close after a permanent error (if ever). This is a design avenue I'd still like to explore for Preview3.

As you noted, there are indeed valid use cases for receiving these transient errors. It just felt too niche to make it into the first version of the proposal.

I think it can even be added afterwards, without breaking the current setup, like this:

```rs
/// Returns a notification stream of network errors.
/// Can maybe even be generalized to include OS-specific diagnostic options like IP_RECVERR. (haven't researched this at all)
get-errors(this: tcp-socket) -> stream<error, _>
```

then the libc implementation of `accept` can call `poll-oneof([accept-stream, error-stream])`.

Reply:
I can see the appeal of a design like `get-errors`, but worry it would create awkward new surface area. How would it work if the user polls on just the `error-stream` and not the `accept-stream`? The underlying host APIs don't have any way to report `ECONNABORTED` without actually trying to accept a connection.

However, I can see the logic of deferring this for now. I've filed #22 to track this, and we can proceed with this PR.